packages/protobuf/src/codegenv1/types.ts

```diff
@@ -74,17 +75,21 @@ export type GenDescExtension<
  *
  * @private
  */
-export type GenDescService<RuntimeShape extends GenDescServiceShape> =
-  DescService & brand<RuntimeShape>;
+export type GenDescService<RuntimeShape extends GenDescServiceShape> = Omit<
+  DescService,
+  "method"
+> & {
+  method: { [K in keyof RuntimeShape]: RuntimeShape[K] & DescMethod };
+};
 
 /**
  * @private
  */
 export type GenDescServiceShape = {
   [localName: string]: {
-    kind: "unary" | "server_streaming" | "client_streaming" | "bidi_streaming";
-    I: DescMessage;
-    O: DescMessage;
+    input: DescMessage;
+    output: DescMessage;
+    methodKind: DescMethod["methodKind"];
```

Comment:
Since `GenDescServiceShape` really only defines methods, WDYT about renaming it, for example:

```ts
export type GenDescServiceMethods = Record<string, Pick<DescMethod, "input" | "output" | "methodKind">>;
```

BTW: I think we can remove the TODOs on line 83 and 84 in packages/protobuf/src/types.ts now.